selfdrive/boardd/panda.cc

```diff
@@ -84,97 +86,70 @@ void Panda::cleanup(){
+  const uint8_t bmRequestType = dir | LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_DEVICE;
+  int r = -1, retries = 0;
+  while (connected && r < 0) {
+    // Release the lock before retrying on failure, so that other threads have a chance to do transfer
+    std::lock_guard lk(usb_lock);
+    r = libusb_control_transfer(dev_handle, bmRequestType, bRequest, wValue, wIndex, NULL, 0, timeout);
+    if (r < 0) handle_usb_issue(r, ++retries, __func__);
   }
+  return r;
+}
 
-  std::lock_guard lk(usb_lock);
-  do {
-    err = libusb_control_transfer(dev_handle, bmRequestType, bRequest, wValue, wIndex, NULL, 0, timeout);
-    if (err < 0) handle_usb_issue(err, __func__);
-  } while (err < 0 && connected);
-
-  return err;
+int Panda::usb_write(uint8_t bRequest, uint16_t wValue, uint16_t wIndex, unsigned int timeout) {
+  return usb_transfer(LIBUSB_ENDPOINT_OUT, bRequest, wValue, wIndex, timeout);
 }
 
 int Panda::usb_read(uint8_t bRequest, uint16_t wValue, uint16_t wIndex, unsigned char *data, uint16_t wLength, unsigned int timeout) {
-  int err;
-  const uint8_t bmRequestType = LIBUSB_ENDPOINT_IN | LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_DEVICE;
-
-  if (!connected){
-    return LIBUSB_ERROR_NO_DEVICE;
-  }
-
-  std::lock_guard lk(usb_lock);
-  do {
-    err = libusb_control_transfer(dev_handle, bmRequestType, bRequest, wValue, wIndex, data, wLength, timeout);
-    if (err < 0) handle_usb_issue(err, __func__);
-  } while (err < 0 && connected);
-
-  return err;
+  return usb_transfer(LIBUSB_ENDPOINT_IN, bRequest, wValue, wIndex, timeout);
 }
 
-int Panda::usb_bulk_write(unsigned char endpoint, unsigned char* data, int length, unsigned int timeout) {
-  int err;
-  int transferred = 0;
-
-  if (!connected){
-    return 0;
-  }
-
+std::tuple<int, int> Panda::usb_bulk_transfer(libusb_endpoint_direction dir, unsigned char endpoint, unsigned char* data, int length, unsigned int timeout) {
+  int err = 0, retries = 0, transferred = 0;
+  // Keep the lock while retrying for bulk transfer
   std::lock_guard lk(usb_lock);
-  do {
-    // Try sending can messages. If the receive buffer on the panda is full it will NAK
-    // and libusb will try again. After 5ms, it will time out. We will drop the messages.
-    err = libusb_bulk_transfer(dev_handle, endpoint, data, length, &transferred, timeout);
-
-    if (err == LIBUSB_ERROR_TIMEOUT) {
-      LOGW("Transmit buffer full");
+  while (connected) {
+    err = libusb_bulk_transfer(dev_handle, dir | endpoint, data, length, &transferred, timeout);
+    if ((err == 0 && (dir != LIBUSB_ENDPOINT_OUT || length == transferred)) || err == LIBUSB_ERROR_TIMEOUT) {
       break;
-    } else if (err != 0 || length != transferred) {
-      handle_usb_issue(err, __func__);
     }
-  } while(err != 0 && connected);
+    if(err == LIBUSB_ERROR_PIPE)
+      libusb_clear_halt(dev_handle, dir | endpoint);
```

Comment:
libusb_bulk_transfer will return LIBUSB_ERROR_PIPE if the endpoint halted, call libusb_clear_halt to clear it before retry